feat: add bulk rename

harryfrzz opened the pull request:
<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Bulk rename UI with modes (find/replace, prefix/suffix, numbering, case conversion, editor mode), live previews, validation, and apply flow
  * Hotkeys to trigger and navigate bulk-rename (B, Tab, Shift+Tab)

* **Documentation**
  * Hotkey docs updated to include bulk rename

* **Tests**
  * Comprehensive tests for transformations, previews, validation, and end-to-end rename operation
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

lazysegtree commented:
@harryfrzz I meant asking yorukot for reviewing feature design.

harryfrzz commented:
@yorukot please review the new **bulk rename** feature that is implemented

lazysegtree commented on src/internal/bulk_rename.go:
```diff
@@ -0,0 +1,169 @@
+package internal
```
> could you please clarify what you meant by “the UI is not stable”?

As I clarified in this comment - https://github.com/yorukot/superfile/pull/1113/files#r2425168810

The empty space between text and border expands while scrolling.

lazysegtree replied:
Let me recheck after your latest changes.

lazysegtree replied:
https://www.loom.com/share/02028689226d412cba150bcc34d6b6cc

lazysegtree commented on src/internal/common/style_function.go:
```diff
@@ -326,3 +326,18 @@ func GenerateFooterBorder(countString string, width int) string {
+	ti.PlaceholderStyle = ModalStyle
+	ti.CharLimit = 156
+	ti.Width = 80 - 12
+	return ti
```
This is correct flag @harryfrzz . Check how other models like prompt manages their searchbar input width. Its not hardcoded anywhere.

I would recommend that you address all the coderabbit flags. Most of them are usually correct.

lazysegtree commented on src/superfile_config/hotkeys.toml:
```diff
@@ -59,3 +59,7 @@ search_bar = ['/', '']
+
+bulk_rename = ['B', '']
+nav_bulk_rename = ['tab','']
+rev_nav_bulk_rename = ['shift+tab','']
```
Also update the vim hotkeys, and hotkey-list in the website docs.

lazysegtree replied:
@coderabbitai Please clarify the location of the files, and point out if anything else is needed to be udpated

lazysegtree commented on src/internal/ui/bulk_rename/render.go:
```diff
@@ -0,0 +1,259 @@
+		return a
+	}
+	return b
+}
```
What is this ? Why are you redefining the existing builtin min functions ?

lazysegtree replied:
Are you using LLM to generate all this code? If so, please ensure that the `slop` is cleaned up.

harryfrzz replied:
> Are you using LLM to generate all this code? If so, please ensure that the `slop` is cleaned up.

Yes I've used LLMs to generate some parts of the code, like to fix codescene errors & code complexity issues etc.
I apologise for not properly maintaining the code.I'll refactor it asap and get back to you

lazysegtree replied:
Ignore the codescene

lazysegtree commented:
@harryfrzz unit tests ?

lazysegtree commented on src/internal/type.go:
```diff
@@ -135,7 +137,72 @@ type typingModal struct {
+	sidebarSearch   bool
+	sortOpen        bool
+	helpOpen        bool
+}
```
Why is this still not fixed @harryfrzz ? Please dont ask for review till you have fixed all the comments.

lazysegtree commented on src/internal/ui/bulk_rename/model.go:
```diff
@@ -0,0 +1,549 @@
+
+	m.preview = previews
+	return previews
+}
```
Please ensure all coderabbit comments are addressed.

harryfrzz commented:
@coderabbitai can you check if the tests for bulk rename is working properly now?

harryfrzz commented:
@coderabbitai check the code that i pushed right now, has it fixed the performance issues and other issues that you have mentioned to me earlier?

harryfrzz commented:
@lazysegtree Thanks for the earlier feedback. I’ve addressed the main points:
- Fixed performance when selecting 1k–10k files.
- Resolved UI spacing/alignment issues; modal width/height are now consistent and centered.
- Added Editor Mode with tmpfile flow and platform editor support.
- Have done tests too, all the tests came out with result PASS
If you find out any issues in the code, please give me a heads up

lazysegtree commented on src/internal/handle_file_operations.go:
```diff
@@ -72,9 +72,6 @@ func (m *model) warnModalForRenaming() tea.Cmd {
 
-// Rename file where the cusror is located
-// TODO: Fix this. It doesn't do any rename, just opens the rename text input
-// Actual rename happens at confirmRename() in handle_modal.go
```
M2 - Please don't remove this comment

lazysegtree commented on src/internal/key_function.go:
M1 - Need to check if the refactoring broke anything

lazysegtree commented on src/internal/model.go:
```diff
@@ -493,6 +575,130 @@ func (m *model) applyZoxideModalAction(action common.ModelAction) {
+
+	m.bulkRenameModel.Close()
+	return bulkrename.ExecuteBulkRename(&m.processBarModel, validPreviews)
+}
```
@coderabbitai - This is a problem with many of our operations. They silently fail and get logged in case of errors and dont show anything to the user.
(1) Find more occurences like this
(2) Create a github issue tracking this - explain why its bad to silently fail and that we should display error notification for important critical failures in file operations.

lazysegtree commented on src/superfile_config/hotkeys.toml:
```diff
@@ -1,5 +1,8 @@
 # =================================================================================================
 # Global hotkeys (cannot conflict with other hotkeys)
+
+# Need to add vim hotkeys for bulk_rename
```
M3 - Please remove this comment

lazysegtree commented on src/internal/ui/bulk_rename/model.go:
```diff
@@ -0,0 +1,565 @@
+	return converted + ext
+}
+
+func toTitleCase(text string) string {
```
M2 - https://github.com/yorukot/superfile/pull/1113/files#r2447881280 This comment is valid.
The functions wont work accurately for multi space strings.

lazysegtree commented on src/internal/ui/bulk_rename/model.go:
M1 - Review the implementation in more detail

lazysegtree commented:
@harryfrzz Please check M2,M3 comments. You can ignore M1 if they are too hard or not clear. You may skip a few of M2 comment too, but please fix most of them. Please fix all M3 comments.

I will then fix everything, do a thorough review and merge this.

lazysegtree commented:
M3 - https://github.com/yorukot/superfile/actions/runs/18993554351/job/54317247938?pr=1113

Fix the gofmt failures

<img width="706" height="442" alt="image" src="https://github.com/user-attachments/assets/f55cd789-6515-416f-94ff-674a43819d2b" />


Also run

`golangci-lint --fix` at the root of the code to fix linting issues.

lazysegtree commented:
Okay. I might take up the rest of the work and finish this if I get time this month.

harryfrzz commented:
> Okay. I might take up the rest of the work and finish this if I get time this month.

I'll fix some of the issues when i get time and sorry for the inconvenience made

harryfrzz commented:
Issues fixed so far @lazysegtree  :
- M2 - Move navigation functions to navigation.go in same package
- M2 - Move all types in this file to a new `type.go` file in the same package
- M3 - Please remove this comment (Removal of comment in `hotkeys.toml`)
- M2 - Values are hardcoded. Its better that default values are moved to constants.
- M3 - [https://github.com/yorukot/superfile/actions/runs/18993554351/job/54317247938?pr=1113](https://github.com/yorukot/superfile/actions/runs/18993554351/job/54317247938?pr=1113) Fix the gofmt failures Also run `golangci-lint --fix` at the root of the code to fix linting issues.
- M3 - Please fix this. Why is 29 here ? (Accidentally added this value)

lazysegtree commented:
I attempted to start fixing the code, but there is just too much to fix. I will check it sometime later.

harryfrzz commented:
> @harryfrzz Meanwhile you should attempt to fix the issues that you can. Many issues are very troublesome for me to fix. I have to go through and understand the entire PR and changes.

@lazysegtree I’ve fixed the remaining bugs in the code. Could you please review it?

harryfrzz commented:
> @harryfrzz Is that okay? I will wait one week for you to respond, else @0xjuicebox is allowed to pick it up.

Yeah sure! My schedule is actually pretty packed right now, so feel free to let @0xjuicebox work on this
